Added model monitoring solutions notebook

kylesteckler opened the pull request:
Solutions notebook for vertex model monitoring lab. Contains solutions/*.ipynb and a pre-trained model in a models folder. 

No updates/edits to any existing content.

BenoitDherin commented on notebooks/model_monitoring/solutions/model_monitoring_vertex.ipynb:
```diff
@@ -0,0 +1,1396 @@
+{
```
<ul><li>Since the part of the notebook relative to deploying the model is just some necessary setup for model monitory, what about packaging the deployment code in a python script using the preferred Python SDK and just invoking that script at the very beginning of the notebook? </li></ul><p> </p><ul><li>Same comment as above relative to <code>beta</code></li><li>I'd add blank lines between the command for readability </li><li>Is the <code>traffic-split</code> argument necessary? If not, remove it, otherwise it deserve some explanation </li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

BenoitDherin replied:
<ul><li>Maybe a concrete example for point 1 would be useful. (e.g.: "If set x to y this means that v percent of the request will be logged" or something like this explaining what the logging frequency means)</li><li>I am not sure I understand point 2</li><li>Maybe some more explanation for point 4 and point 5 are needed</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

BenoitDherin replied:
<ul><li>What is happening in this cell is very opaque. I think the code here needs to be broken down into smaller chunks and explained in markdown cells. I also suggest to completely get rid of all the helper functions and use the API directly, explaining what each call does and why.</li></ul><p> </p><ul><li>This cell failed for me, so I was not able to fully test the notebook to the end.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

BenoitDherin replied:
<ul><li>"receive email" -&gt; "receive an email"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

BenoitDherin replied:
<ul><li>"This is the first prediction made": Maybe explain why we need to record only the first prediction (seems weird and not obvious to me)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

BenoitDherin replied:
<p>I'd remove that, or put it as a "<strong>remark:" </strong>somewhere. As it is it seems the corresponding concrete explanation is missing. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

BenoitDherin replied:
<p>Maybe explain which threshold will be triggered and how by the next cell to illustrate what monitoring capability we are demonstrating.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

BenoitDherin replied:
<p>This function is not very easy to read. Is there a way to simplify it, maybe by breaking it into smaller functions with clear names describing what they do. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

BenoitDherin replied:
<p>Maybe explain what the Jensen-Shannon divergence and how it is used in this context. Same thing with "High Linfty". Information regarding when the metric values can be considered as significant or not may be also helpful. Right now, it's not super what to do with this output without a little more guidance. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

BenoitDherin replied:
<p>It's not very clear how to infer that conclusion from the tool output. Maybe some more explanation of the meaning of the metrics used may help. Providing a list of the main anomalies and their explanation may also be useful. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

BenoitDherin replied:
<p>Add a link to the monitoring console.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

BenoitDherin commented on notebooks/model_monitoring/solutions/model_monitoring_vertex.ipynb:
```diff
@@ -0,0 +1,1396 @@
+{
```
<p>Some explanation may be useful here regarding what info to extract from these screenshots. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

sanjanalreddy replied:
<p>+1 This is important</p>

kylesteckler replied:
<p>Done.</p>

BenoitDherin reviewed (changes requested):
Added detailed comments using the NBReview tool. 

Main points:

1) Consider removing all the helper functions and replace them by explicit calls to the API, creating the opportunity to explain (in markdown cells) what these calls are doing and why they are needed. (This may also help simplify the monitoring tool setup.)

2) Consider packaging the code to deploy the canned model into its on script (using the Python SDK) so that it doesn't take so much space in the notebook, leaving real estate to give more explanations about the monitoring system.

3) Consider adding more explanations around the monitoring tool output so that it is clearer how to react to its alerts.

Otherwise, great lab! Thanks Kyle for creating it!

kylesteckler commented:
Major changes in new commit based on code review:
 - Entirely got rid of helper functions. Refactored functionality with python client library and/or gcloud ai 
 - Split up monitoring job create section into multiple cell blocks with deeper explanations
 - Links to concepts related to explanations 
 - Minor syntax/wording fixes

sanjanalreddy commented on notebooks/model_monitoring/solutions/model_monitoring_vertex.ipynb:
```diff
@@ -0,0 +1,1294 @@
+{
```
<p>We can remove this header and make the next one a bigger header, similar to other notebooks.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

kylesteckler replied:
<p>Done.</p>

sanjanalreddy commented on notebooks/model_monitoring/solutions/model_monitoring_vertex.ipynb:
```diff
@@ -0,0 +1,1294 @@
+{
```
<p>Encapsulate this in an if statement, so the code only tries to create a bucket if it doesn't exist. This is cleaner.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

kylesteckler replied:
<p>Done.</p>

sanjanalreddy commented on notebooks/model_monitoring/solutions/model_monitoring_vertex.ipynb:
```diff
@@ -0,0 +1,1294 @@
+{
```
<p>Can we define L-inifnity distance and Jensen-Shannon Divergence in a few words? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

BenoitDherin replied:
<p>+1 </p><p> </p><p>I also think we need brief definitions here. </p>

kylesteckler replied:
<p>Done. </p>

sanjanalreddy commented on notebooks/model_monitoring/solutions/model_monitoring_vertex.ipynb:
```diff
@@ -0,0 +1,1294 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #8.        input_uri="bq://mco-mm.bqmlga4.train"
</pre><p>Before we merge, let's try to replace this with our own BQ dataset or use CSV files. This will make us completely independent from someone else's BQ dataset.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

kylesteckler replied:
<p>Done.</p>

sanjanalreddy commented on notebooks/model_monitoring/solutions/model_monitoring_vertex.ipynb:
```diff
@@ -0,0 +1,1294 @@
+{
```
<p>Consider explaining that the minimum amount of time that you have to wait for an alert is one hour and it can't be changed to less as of now. I think customers would like it to be sooner but unfortunately, Vertex does not let you configure it to be less than an hour.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/asl-ml-immersion/pull/209/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

sanjanalreddy reviewed (changes requested):
I suggested a few changes. But I would like to hold on merging this into master till we change the BQ dataset to read from an ASL dataset or use CSV files and then upload to BQ. The dataset you are currently using is actively being used by DevRel and it might lead to unintentional changes if we don't do a copy now.

BenoitDherin reviewed (approved):
LGTM. Super clear. Great Lab!

BenoitDherin commented:
(Lab ran perfectly for me.)

takumiohym reviewed (approved):
LGTM!
